src/frequency_cam_ros2.cpp

```diff
@@ -150,6 +182,71 @@ void FrequencyCamROS::playEventsFromBag(const std::string & bagName, const std::
   rclcpp::shutdown();
 }
 
+void FrequencyCamROS::playEventsFromBagFrameBased(
+  const std::string & bagName, const std::string & bagTopic, const std::string & triggers_file)
+{
+  imageMaker_.setScale(this->declare_parameter<double>("scale_image", 1.0));
+  rclcpp::Time lastFrameTime(0);
+  rosbag2_cpp::Reader reader;
+  reader.open(bagName);
+  rclcpp::Serialization<EventArray> serialization;
+  const auto delta_t = rclcpp::Duration::from_seconds(eventImageDt_);
+  uint32_t frameCount(0);
+  const std::string path = this->declare_parameter<std::string>("path", "./frames");
+  std::filesystem::create_directories(path);
+
+  setTriggers(triggers_file);
+
+  uint64_t timeLimit = 0;
+
+  std::size_t triggerIndex = 0;
+  while (reader.has_next()) {
+    auto bagmsg = reader.read_next();
+    rclcpp::SerializedMessage serializedMsg(*bagmsg->serialized_data);
+    if (bagmsg->topic_name != bagTopic) {
+      continue;
+    }
+    EventArray::SharedPtr msg(new EventArray());
+    serialization.deserialize_message(&serializedMsg, &(*msg));
+    if (msg) {
+      const rclcpp::Time t(msg->header.stamp);
+      eventMsg(msg);
+      auto decoder = decoderFactory_.getInstance(msg->encoding, msg->width, msg->height);
+      timeLimit = externalTriggers_.at(triggerIndex++);
+      size_t numConsumed = decoder->decodeUntil(
+        msg->events.data(), msg->events.size(), &cam_, timeLimit, nullptr /*uint64_t * nextTime*/);
+
+      if (numConsumed < msg->events.size()) {
+        std::cerr << "There are more events to be processed" << std::endl;
```

Comment:
This is a problem. You really need to handle the fractional packets.
Please have a look at my example implementation in the "frame_based" branch I just added.

Reply:
Sorry, this is very much work in progress code.  I planned to ping you once I believe I have a working implementation. But thanks for your example, I will take a look at it.